File: streampark-common/src/main/scala/org/apache/streampark/common/conf/FlinkVersion.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.common.conf

import org.apache.streampark.common.util.{CommandUtils, Logger}
import org.apache.streampark.common.util.Implicits._

import java.io.File
import java.net.URL
import java.util.function.Consumer
import java.util.regex.Pattern

import scala.collection.mutable

/** @param flinkHome actual flink home that must be a readable local path */
class FlinkVersion(val flinkHome: String) extends Serializable with Logger {

  private[this] lazy val FLINK_VER_PATTERN = Pattern.compile("^(\\d+\\.\\d+)(\\.)?.*$")

  private[this] lazy val FLINK_VERSION_PATTERN = Pattern.compile("^Version: (.*), Commit ID: (.*)$")

  private[this] lazy val FLINK_SCALA_VERSION_PATTERN =
    Pattern.compile("^flink-dist_(\\d\\.\\d+).*.jar$")

  private[this] lazy val APACHE_FLINK_VERSION_PATTERN = Pattern.compile("(^\\d+\\.\\d+\\.\\d+)")

  private[this] lazy val OTHER_FLINK_VERSION_PATTERN = Pattern.compile("(\\d+\\.\\d+)(-*)")

  lazy val scalaVersion: String = {
    val matcher = FLINK_SCALA_VERSION_PATTERN.matcher(flinkDistJar.getName)
    if (matcher.matches()) {
      matcher.group(1)
    } else {
      // flink 1.15 + on support scala 2.12
      "2.12"
    }
  }

  lazy val fullVersion: String = s"${version}_$scalaVersion"

  lazy val flinkLib: File = {
    require(flinkHome != null, "[StreamPark] flinkHome must not be null.")
    require(new File(flinkHome).exists(), "[StreamPark] flinkHome must be exists.")
    val lib = new File(s"$flinkHome/lib")
    require(
      lib.exists() && lib.isDirectory,
      s"[StreamPark] $flinkHome/lib must be exists and must be directory.")
    lib
  }

  lazy val flinkLibs: List[URL] = flinkLib.listFiles().map(_.toURI.toURL).toList

  lazy val version: String = {
    val cmd = List(
      s"java -classpath ${flinkDistJar.getName} org.apache.flink.client.cli.CliFrontend --version")
    var flinkVersion: String = null
    val buffer = new mutable.StringBuilder
    CommandUtils.execute(
      flinkLib.getAbsolutePath,
      cmd,
      new Consumer[String]() {
        override def accept(out: String): Unit = {
          buffer.append(out).append("\n")
          val matcher = FLINK_VERSION_PATTERN.matcher(out)
          if (matcher.find) {
            val version = matcher.group(1)
            val matcher1 = APACHE_FLINK_VERSION_PATTERN.matcher(version)
            if (matcher1.find) {
              flinkVersion = version
            } else {
              val matcher2 = OTHER_FLINK_VERSION_PATTERN.matcher(version)
              if (matcher2.find) {
                flinkVersion = version
              }
            }
          }
        }
      })

    logInfo(buffer.toString())
    if (flinkVersion == null) {
      throw new IllegalStateException(s"[StreamPark] parse flink version failed. $buffer")
    }
    buffer.clear()
    flinkVersion
  }

  // flink major version, like "1.13", "1.14"
  lazy val majorVersion: String = {
    if (version == null) {
      null
    } else {
      val matcher = FLINK_VER_PATTERN.matcher(version)
      matcher.matches()
      matcher.group(1)
    }
  }

  lazy val flinkDistJar: File = {
    val distJar = flinkLib.listFiles().filter(_.getName.matches("flink-dist.*\\.jar"))
    distJar match {
      case x if x.isEmpty =>
        throw new IllegalArgumentException(s"[StreamPark] can no found flink-dist jar in $flinkLib")
      case x if x.length > 1 =>
        throw new IllegalArgumentException(
          s"[StreamPark] found multiple flink-dist jar in $flinkLib")
      case _ =>
    }
    distJar.head
  }

  def checkVersion(throwException: Boolean = true): Boolean = {
    version.split("\\.").map(_.trim.toInt) match {
      case Array(1, v, _) if v >= 12 && v <= 19 => true
      case _ =>
        if (throwException) {
          throw new UnsupportedOperationException(s"Unsupported flink version: $version")
        } else {
          false
        }
    }
  }

  def checkVersion(sinceVersion: Int): Boolean = {
    version.split("\\.").map(_.trim.toInt) match {
      case Array(1, v, _) if v >= sinceVersion => true
      case _ => false
    }
  }

  // StreamPark flink shims version, like "streampark-flink-shims_flink-1.13"
  private lazy val shimsVersion: String = s"streampark-flink-shims_flink-$majorVersion"

  override def toString: String =
    s"""
       |----------------------------------------- flink version -----------------------------------
       |     flinkHome    : $flinkHome
       |     distJarName  : ${flinkDistJar.getName}
       |     flinkVersion : $version
       |     majorVersion : $majorVersion
       |     scalaVersion : $scalaVersion
       |     shimsVersion : $shimsVersion
       |-------------------------------------------------------------------------------------------
       |""".stripMargin

}

File: streampark-common/src/main/java/org/apache/streampark/common/conf/FlinkVersion.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.common.conf;

import org.apache.streampark.common.util.CommandUtils;

import org.apache.streampark.shaded.org.slf4j.Logger;

import java.io.File;
import java.io.Serializable;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.function.Consumer;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

/** @param flinkHome actual flink home that must be a readable local path */
public class FlinkVersion implements Serializable {

    private static final long serialVersionUID = 1L;
    private static final Logger LOG =
        org.apache.streampark.common.util.StreamParkLoggerFactory.loggerFactory()
            .getLogger(FlinkVersion.class.getName());

    private static final Pattern FLINK_VER_PATTERN = Pattern.compile("^(\\d+\\.\\d+)(?:\\..*)?$");
    private static final Pattern FLINK_VERSION_PATTERN =
        Pattern.compile("^Version: ([^,]*), Commit ID: (.*)$");
    private static final Pattern FLINK_DIST_UNDERSCORE_PATTERN =
        Pattern.compile(
            "^flink-dist_(\\d+\\.\\d+)-(\\d+\\.\\d+(?:\\.\\d+)?(?:-SNAPSHOT)?)\\.jar$");
    private static final Pattern FLINK_DIST_DASH_PATTERN =
        Pattern.compile("^flink-dist-(\\d+\\.\\d+(?:\\.\\d+)?(?:-SNAPSHOT)?)\\.jar$");
    private static final Pattern APACHE_FLINK_VERSION_PATTERN =
        Pattern.compile("^(\\d+\\.\\d+\\.\\d+)");
    private static final Pattern OTHER_FLINK_VERSION_PATTERN = Pattern.compile("^(\\d+\\.\\d+)-?$");

    /** Flink installation directory (Scala {@code flinkHome} accessor). */
    public final String flinkHome;
    private transient String scalaVersion;
    private transient String version;
    private transient File flinkLib;
    private transient File flinkDistJar;

    public FlinkVersion(String flinkHome) {
        this.flinkHome = flinkHome;
    }

    public String getFlinkHome() {
        return flinkHome;
    }

    /** Scala API alias for {@link #getFullVersion()}. */
    public String fullVersion() {
        return getFullVersion();
    }

    /** Scala API alias for {@link #getMajorVersion()}. */
    public String majorVersion() {
        return getMajorVersion();
    }

    /** Scala API alias for {@link #getScalaVersion()}. */
    public String scalaVersion() {
        return getScalaVersion();
    }

    /** Scala API alias for {@link #getVersion()}. */
    public String version() {
        return getVersion();
    }

    /** Backward-compatible alias for {@link #getFlinkLibs()}. */
    public List<URL> flinkLibs() throws Exception {
        return getFlinkLibs();
    }

    public String getScalaVersion() {
        if (scalaVersion == null) {
            scalaVersion = parseFromDistJar().map(pair -> pair[1]).orElse("2.12");
        }
        return scalaVersion;
    }

    public String getFullVersion() {
        return getVersion() + "_" + getScalaVersion();
    }

    public File getFlinkLib() {
        if (flinkLib == null) {
            if (flinkHome == null) {
                throw new IllegalArgumentException("[StreamPark] flinkHome must not be null.");
            }
            if (!new File(flinkHome).exists()) {
                throw new IllegalArgumentException("[StreamPark] flinkHome must be exists.");
            }
            File lib = new File(flinkHome + "/lib");
            if (!lib.exists() || !lib.isDirectory()) {
                throw new IllegalArgumentException(
                    "[StreamPark] " + flinkHome + "/lib must be exists and must be directory.");
            }
            flinkLib = lib;
        }
        return flinkLib;
    }

    public List<URL> getFlinkLibs() throws Exception {
        File[] files = getFlinkLib().listFiles();
        if (files == null) {
            return Collections.emptyList();
        }
        return Arrays.stream(files).map(f -> {
            try {
                return f.toURI().toURL();
            } catch (MalformedURLException e) {
                throw new IllegalArgumentException("Invalid Flink lib URL: " + f, e);
            }
        }).collect(Collectors.toList());
    }

    public String getVersion() {
        if (version == null) {
            version =
                parseFromDistJar()
                    .map(pair -> pair[0])
                    .orElseGet(this::parseFromCliFrontend);
        }
        return version;
    }

    public String getMajorVersion() {
        if (getVersion() == null) {
            return null;
        }
        Matcher matcher = FLINK_VER_PATTERN.matcher(getVersion());
        matcher.matches();
        return matcher.group(1);
    }

    public File getFlinkDistJar() {
        if (flinkDistJar == null) {
            File[] distJar =
                getFlinkLib().listFiles(
                    f -> {
                        String name = f.getName();
                        return name.startsWith("flink-dist") && name.endsWith(".jar");
                    });
            if (distJar == null || distJar.length == 0) {
                throw new IllegalArgumentException(
                    "[StreamPark] can no found flink-dist jar in " + getFlinkLib());
            }
            if (distJar.length > 1) {
                throw new IllegalArgumentException(
                    "[StreamPark] found multiple flink-dist jar in " + getFlinkLib());
            }
            flinkDistJar = distJar[0];
        }
        return flinkDistJar;
    }

    public boolean checkVersion() {
        return checkVersion(true);
    }

    public boolean checkVersion(boolean throwException) {
        String[] parts = getVersion().split("\\.");
        if (parts.length >= 2) {
            try {
                int major = Integer.parseInt(parts[0].trim());
                int minor = Integer.parseInt(parts[1].trim());
                if (major == 1 && minor >= 17 && minor <= 20) {
                    return true;
                }
                if (major == 2 && minor >= 0 && minor <= 3) {
                    return true;
                }
            } catch (NumberFormatException ignored) {
            }
        }
        if (throwException) {
            throw new UnsupportedOperationException("Unsupported flink version: " + getVersion());
        }
        return false;
    }

    public boolean checkVersion(int sinceVersion) {
        String[] parts = getVersion().split("\\.");
        if (parts.length >= 2) {
            try {
                int major = Integer.parseInt(parts[0].trim());
                int minor = Integer.parseInt(parts[1].trim());
                if (major == 1 && minor >= sinceVersion) {
                    return true;
                }
                if (major == 2) {
                    return true;
                }
            } catch (NumberFormatException ignored) {
            }
        }
        return false;
    }

    private java.util.Optional<String[]> parseFromDistJar() {
        String jarName = getFlinkDistJar().getName();
        Matcher underscoreMatcher = FLINK_DIST_UNDERSCORE_PATTERN.matcher(jarName);
        if (underscoreMatcher.matches()) {
            String parsedVersion = underscoreMatcher.group(2);
            String parsedScala = underscoreMatcher.group(1);
            LOG.info(
                "Flink version parsed from dist jar name: {}, scala: {}",
                parsedVersion,
                parsedScala);
            return java.util.Optional.of(new String[]{parsedVersion, parsedScala});
        }
        Matcher dashMatcher = FLINK_DIST_DASH_PATTERN.matcher(jarName);
        if (dashMatcher.matches()) {
            String parsedVersion = dashMatcher.group(1);
            LOG.info(
                "Flink version parsed from dist jar name: {}, scala: {}",
                parsedVersion,
                "2.12");
            return java.util.Optional.of(new String[]{parsedVersion, "2.12"});
        }
        return java.util.Optional.empty();
    }

    private String parseFromCliFrontend() {
        final String[] flinkVersion = {null};
        StringBuilder buffer = new StringBuilder();
        List<String> cmd =
            Arrays.asList(
                "java -classpath "
                    + getFlinkDistJar().getName()
                    + " org.apache.flink.client.cli.CliFrontend --version");
        try {
            CommandUtils.execute(
                getFlinkLib().getAbsolutePath(),
                cmd,
                new Consumer<String>() {

                    @Override
                    public void accept(String out) {
                        buffer.append(out).append("\n");
                        Matcher matcher = FLINK_VERSION_PATTERN.matcher(out);
                        if (matcher.find()) {
                            String ver = matcher.group(1);
                            Matcher m1 = APACHE_FLINK_VERSION_PATTERN.matcher(ver);
                            if (m1.find()) {
                                flinkVersion[0] = ver;
                            } else {
                                Matcher m2 = OTHER_FLINK_VERSION_PATTERN.matcher(ver);
                                if (m2.find()) {
                                    flinkVersion[0] = ver;
                                }
                            }
                        }
                    }
                });
        } catch (Exception e) {
            throw new IllegalStateException("[StreamPark] execute flink version command failed", e);
        }
        LOG.info("[StreamPark] {}", buffer);
        if (flinkVersion[0] == null) {
            throw new IllegalStateException("[StreamPark] parse flink version failed. " + buffer);
        }
        return flinkVersion[0];
    }

    @Override
    public String toString() {
        return "\n----------------------------------------- flink version -----------------------------------\n"
            + "     flinkHome    : "
            + flinkHome
            + "\n     distJarName  : "
            + getFlinkDistJar().getName()
            + "\n     flinkVersion : "
            + getVersion()
            + "\n     majorVersion : "
            + getMajorVersion()
            + "\n     scalaVersion : "
            + getScalaVersion()
            + "\n     shimsVersion : streampark-flink-shims_flink-"
            + getMajorVersion()
            + "\n-------------------------------------------------------------------------------------------\n";
    }
}
